File: packages/utils/src/redact.ts
const redact = (
  text: string,
  replacementChar = "*",
  revealChars = 4,
  direction: "start" | "end" = "start",
) => {
  if (!text) return "";
  const visible = Math.max(0, revealChars);
  if (text.length <= visible) return text;
  const redacted = (replacementChar[0] ?? "*").repeat(text.length - visible);
  if (direction === "start") return `${text.slice(0, visible)}${redacted}`;
  return `${redacted}${text.slice(-visible)}`;
};

export { redact };

File: apps/ingest/pkg/util/redact.go
package util

import "strings"

const (
	RedactDirectionStart = "start"
	RedactDirectionEnd   = "end"
)

func Redact(text string, revealChars int, direction string) string {
	if text == "" {
		return ""
	}

	visible := max(0, revealChars)
	if len(text) <= visible {
		return text
	}

	redacted := strings.Repeat("*", len(text)-visible)
	if direction == RedactDirectionStart {
		return text[:visible] + redacted
	}

	return redacted + text[len(text)-visible:]
}
